### ISSUE TITLE
jekyll build can't find jekyll-gist

### ISSUE THREAD

**Issue by toogley** [NixOS/nixpkgs#22238]
I want to use the program `jekyll` to build some websites, but i don't know how i can make jekyll find other libraries.

```
$ nix-env -iA nixos.ruby_2_2
$ nix-env -i '.*jekyll.*'
$ gem install jekyll-gist  jekyll-sitemap jekyll-paginate
$ jekyll build
Configuration file: /home/toogley/src/neomutt.github.io/_config.yml
/nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/jekyll-3.0.1/lib/jekyll/plugin_manager.rb:30:in `require': cannot load such file -- jekyll-gist (LoadError)
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/jekyll-3.0.1/lib/jekyll/plugin_manager.rb:30:in `block in require_gems'
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/jekyll-3.0.1/lib/jekyll/plugin_manager.rb:27:in `each'
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/jekyll-3.0.1/lib/jekyll/plugin_manager.rb:27:in `require_gems'
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/jekyll-3.0.1/lib/jekyll/plugin_manager.rb:19:in `conscientious_require'
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/jekyll-3.0.1/lib/jekyll/site.rb:97:in `setup'
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/jekyll-3.0.1/lib/jekyll/site.rb:49:in `initialize'
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/jekyll-3.0.1/lib/jekyll/commands/build.rb:30:in `new'
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/jekyll-3.0.1/lib/jekyll/commands/build.rb:30:in `process'
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/jekyll-3.0.1/lib/jekyll/commands/build.rb:18:in `block (2 levels) in init_with_program'
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `call'
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `block in execute'
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `each'
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `execute'
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/mercenary-0.3.5/lib/mercenary/program.rb:42:in `go'
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/mercenary-0.3.5/lib/mercenary.rb:19:in `program'
	from /nix/store/fqmibdyxg8pph4wwircdhn1s0ix0bnag-jekyll-3.0.1/lib/ruby/gems/2.2.0/gems/jekyll-3.0.1/bin/jekyll:17:in `<top (required)>'
	from /home/toogley/.nix-profile/bin/jekyll:18:in `load'
	from /home/toogley/.nix-profile/bin/jekyll:18:in `<main>'
```

(using sudo for installing the gems also doesn't help)

## Technical details

* System: `16.09.1587.b186574 (Flounder)`
* Nix version: `nix-env (Nix) 1.11.6`
* Nixpkgs version: `nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion`

**Comment by lheckemann** [comment]
(in case this is still relevant...)

I don't really understand the details of how ruby stuff works, but as far as I can see you'll need to override the jekyll expression to add the appropriate dependencies — take a look at `pkgs/applications/misc/jekyll`.

You may also be able to load gems from your home directory (which I presume is where gem installed those libraries; for me it doesn't even get that far) by setting the GEM_PATH environment variable as appropriate.

https://nixos.org/nixpkgs/manual/#sec-language-ruby might also be helpful.

Sorry for the fairly vague answer, as I said I don't know much about ruby in general and even less about ruby in nix, but hopefully these are at least some useful avenues for exploration...

**Comment by toogley** [comment]
@lheckemann thank you - but i don't know if and when  i'll be test this plan.

**Comment by mmahut** [comment]
Are there any updates to this issue, please?

**Comment by primeos** [comment]
I didn't have a look at this issue specifically but I think this is caused by the same problem as #58126 (the relevant part is in "Fixing the second problem" of https://github.com/NixOS/nixpkgs/issues/58126#issuecomment-475862039).

There should be two possible solutions as roughly described in https://github.com/NixOS/nixpkgs/issues/58126#issuecomment-482269622:
- Patch the Nixpkgs wrapper to avoid setting `BUNDLE_GEMFILE` (might cause other problems and requires changes to multiple files - not trivial and untested, might not even work as expected)
- Patch Jekyll (might be the better solution but I don't know how to apply custom patches to Ruby packages in Nixpkgs, maybe there is no (easy) way to do this - and this solution is also untested).

So to sum it up: AFAIK we have identified the problem and there should be at least two possible solutions but I'm not aware of any easy way to test and apply these fixes (if someone has time to play around with the Ruby infrastructure in Nixpkgs that would be welcome :smile:).

**Comment by primeos** [comment]
I just found this commit: https://github.com/NixOS/nixpkgs/commit/d24aefd52f75f0854a417c3adf02c9d310c23157 which I totally forgot about. Maybe I'll work as a workaround for this issue.

**Comment by raboof** [comment]
I think you should probably install `jekyll.override { withOptionalDependencies = true; }` instead - jekyll-gist appears to be part of that.

**Comment by lheckemann** [comment]
Given @raboof's solution, I think we can close this. Feel free to comment or open a new issue if this is wrong :)
